### PR TITLE
Pull in upstream changes from 0.15.0, take 2

### PR DESCRIPTION
**Pull request by nebrius**
I tried getting https://github.com/observablehq/lezer/pull/7 to work, but couldn't figure out what was going on. I recreated this PR from scratch by pulling in commits directly from the parent repo. Tests now pass:

Screen Shot 2021-08-30 at 5 05 02 PM

### CODE REVIEW

**Comment by nebrius** [comment]
Oh haha, wait. I gotta resolve conflicts first...how much you wanna bet I end up getting back to the same error 😂 

Converting this back to a draft.

**Comment by nebrius** [comment]
Yup, I did indeed get back to that same error

File: src/javascript.grammar [hunk 1]
```diff
@@ -32,8 +32,13 @@
   ExpressionStatement |
   NamedBlockCell[@dynamicPrecedence=1] { (VariableDefinition | QualifiedVariableDefinition) !cellname "=" Block } |
   NamedExpressionCell[@dynamicPrecedence=1] { (VariableDefinition | QualifiedVariableDefinition) !cellname "=" expression semi } |
-  NamedClassCell { kw<"class"> !cellname VariableDefinition (kw<"extends"> expression)? ClassBody } |
-  NamedFunctionCell { async? kw<"function"> Star? !cellname VariableDefinition ParamList Block }
+  // We have to supply rules for both named and unnamed classes and functions
```
Comment (nebrius) [inline src/javascript.grammar:34]: Got tests to pass!

Screen Shot 2021-08-31 at 11 51 25 AM

Turns out our `NamedClassCell` wasn't taking precedence over `ClassDeclaration`, which had `!statement` as it's first precedence marker, which is marked as `@cut` and thus caused it to take precedence.

To fix this, I added `!statement` to our rules too, which forces it to override `ClassDeclaration` if it's a top-level statement. However, this introduced a new issue: top level unnamed expressions were getting assigned to `NamedClassCell`/`NamedFunctionCell` with an error marker (cause of a missing name). To fix this, I added two new rules for unnamed class cells and unnamed class functions.

Comment (sydneypalumbo) [reply src/javascript.grammar:34]: makes sense, thanks for tackling this Bryan!! 🙏 

wonder if we could just get rid of the `cellname` rule at this point -- might be worth a try? not critical, just don't think it's doing anything? (could be wrong though 😅)

**Review by sydneypalumbo — APPROVED** [review]

File: src/javascript.grammar (hunk 1 above)
Comment (nebrius) [reply src/javascript.grammar:34]: Yep, totally doable and tests still pass

**Comment by nebrius** [comment]
FYI I'm going to do a standard merge, instead of a squash merge, so that we can get Marijnh's commits in as-is, to help keep our history trees more in sync.

**Pull request merged**